File: java/lib/src/main/java/com/svix/models/MessageAttemptLog.java
// This file is @generated
package com.svix.models;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.svix.Utils;

import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.time.OffsetDateTime;

@ToString
@EqualsAndHashCode
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonAutoDetect(getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class MessageAttemptLog {
    @JsonProperty private String appId;
    @JsonProperty private String appUid;
    @JsonProperty private Long attemptCount;
    @JsonProperty private OffsetDateTime attemptEnd;
    @JsonProperty private String attemptId;
    @JsonProperty private OffsetDateTime attemptStart;
    @JsonProperty private String endpointId;
    @JsonProperty private String eventType;
    @JsonProperty private HttpAttemptTimes httpTimes;
    @JsonProperty private OffsetDateTime msgCreated;
    @JsonProperty private String msgEventId;
    @JsonProperty private String msgId;
    @JsonProperty private String orgId;
    @JsonProperty private Short responseStatusCode;
    @JsonProperty private MessageStatus status;

    public MessageAttemptLog() {}

    public MessageAttemptLog appId(String appId) {
        this.appId = appId;
        return this;
    }

    /**
     * The Application's ID.
     *
     * @return appId
     */
    @javax.annotation.Nonnull
    public String getAppId() {
        return appId;
    }

    public void setAppId(String appId) {
        this.appId = appId;
    }

    public MessageAttemptLog appUid(String appUid) {
        this.appUid = appUid;
        return this;
    }

    /**
     * The Application's UID.
     *
     * @return appUid
     */
    @javax.annotation.Nullable
    public String getAppUid() {
        return appUid;
    }

    public void setAppUid(String appUid) {
        this.appUid = appUid;
    }

    public MessageAttemptLog attemptCount(Long attemptCount) {
        this.attemptCount = attemptCount;
        return this;
    }

    /**
     * Get attemptCount
     *
     * @return attemptCount
     */
    @javax.annotation.Nonnull
    public Long getAttemptCount() {
        return attemptCount;
    }

    public void setAttemptCount(Long attemptCount) {
        this.attemptCount = attemptCount;
    }

    public MessageAttemptLog attemptEnd(OffsetDateTime attemptEnd) {
        this.attemptEnd = attemptEnd;
        return this;
    }

    /**
     * Get attemptEnd
     *
     * @return attemptEnd
     */
    @javax.annotation.Nonnull
    public OffsetDateTime getAttemptEnd() {
        return attemptEnd;
    }

    public void setAttemptEnd(OffsetDateTime attemptEnd) {
        this.attemptEnd = attemptEnd;
    }

    public MessageAttemptLog attemptId(String attemptId) {
        this.attemptId = attemptId;
        return this;
    }

    /**
     * The MessageAttempt's ID.
     *
     * @return attemptId
     */
    @javax.annotation.Nonnull
    public String getAttemptId() {
        return attemptId;
    }

    public void setAttemptId(String attemptId) {
        this.attemptId = attemptId;
    }

    public MessageAttemptLog attemptStart(OffsetDateTime attemptStart) {
        this.attemptStart = attemptStart;
        return this;
    }

    /**
     * Get attemptStart
     *
     * @return attemptStart
     */
    @javax.annotation.Nonnull
    public OffsetDateTime getAttemptStart() {
        return attemptStart;
    }

    public void setAttemptStart(OffsetDateTime attemptStart) {
        this.attemptStart = attemptStart;
    }

    public MessageAttemptLog endpointId(String endpointId) {
        this.endpointId = endpointId;
        return this;
    }

    /**
     * The Endpoint's ID.
     *
     * @return endpointId
     */
    @javax.annotation.Nonnull
    public String getEndpointId() {
        return endpointId;
    }

    public void setEndpointId(String endpointId) {
        this.endpointId = endpointId;
    }

    public MessageAttemptLog eventType(String eventType) {
        this.eventType = eventType;
        return this;
    }

    /**
     * The event type's name
     *
     * @return eventType
     */
    @javax.annotation.Nullable
    public String getEventType() {
        return eventType;
    }

    public void setEventType(String eventType) {
        this.eventType = eventType;
    }

    public MessageAttemptLog httpTimes(HttpAttemptTimes httpTimes) {
        this.httpTimes = httpTimes;
        return this;
    }

    /**
     * Get httpTimes
     *
     * @return httpTimes
     */
    @javax.annotation.Nullable
    public HttpAttemptTimes getHttpTimes() {
        return httpTimes;
    }

    public void setHttpTimes(HttpAttemptTimes httpTimes) {
        this.httpTimes = httpTimes;
    }

    public MessageAttemptLog msgCreated(OffsetDateTime msgCreated) {
        this.msgCreated = msgCreated;
        return this;
    }

    /**
     * Get msgCreated
     *
     * @return msgCreated
     */
    @javax.annotation.Nonnull
    public OffsetDateTime getMsgCreated() {
        return msgCreated;
    }

    public void setMsgCreated(OffsetDateTime msgCreated) {
        this.msgCreated = msgCreated;
    }

    public MessageAttemptLog msgEventId(String msgEventId) {
        this.msgEventId = msgEventId;
        return this;
    }

    /**
     * The Message's UID.
     *
     * @return msgEventId
     */
    @javax.annotation.Nullable
    public String getMsgEventId() {
        return msgEventId;
    }

    public void setMsgEventId(String msgEventId) {
        this.msgEventId = msgEventId;
    }

    public MessageAttemptLog msgId(String msgId) {
        this.msgId = msgId;
        return this;
    }

    /**
     * The Message's ID.
     *
     * @return msgId
     */
    @javax.annotation.Nonnull
    public String getMsgId() {
        return msgId;
    }

    public void setMsgId(String msgId) {
        this.msgId = msgId;
    }

    public MessageAttemptLog orgId(String orgId) {
        this.orgId = orgId;
        return this;
    }

    /**
     * The Environment's ID.
     *
     * @return orgId
     */
    @javax.annotation.Nonnull
    public String getOrgId() {
        return orgId;
    }

    public void setOrgId(String orgId) {
        this.orgId = orgId;
    }

    public MessageAttemptLog responseStatusCode(Short responseStatusCode) {
        this.responseStatusCode = responseStatusCode;
        return this;
    }

    /**
     * Get responseStatusCode
     *
     * @return responseStatusCode
     */
    @javax.annotation.Nonnull
    public Short getResponseStatusCode() {
        return responseStatusCode;
    }

    public void setResponseStatusCode(Short responseStatusCode) {
        this.responseStatusCode = responseStatusCode;
    }

    public MessageAttemptLog status(MessageStatus status) {
        this.status = status;
        return this;
    }

    /**
     * Get status
     *
     * @return status
     */
    @javax.annotation.Nonnull
    public MessageStatus getStatus() {
        return status;
    }

    public void setStatus(MessageStatus status) {
        this.status = status;
    }

    /**
     * Create an instance of MessageAttemptLog given an JSON string
     *
     * @param jsonString JSON string
     * @return An instance of MessageAttemptLog
     * @throws JsonProcessingException if the JSON string is invalid with respect to
     *     MessageAttemptLog
     */
    public static MessageAttemptLog fromJson(String jsonString) throws JsonProcessingException {
        return Utils.getObjectMapper().readValue(jsonString, MessageAttemptLog.class);
    }

    /**
     * Convert an instance of MessageAttemptLog to an JSON string
     *
     * @return JSON string
     */
    public String toJson() throws JsonProcessingException {
        return Utils.getObjectMapper().writeValueAsString(this);
    }
}

File: php/src/Models/MessageAttemptLog.php
<?php

// this file is @generated
declare(strict_types=1);

namespace Svix\Models;

class MessageAttemptLog implements \JsonSerializable
{
    private array $setFields = [];

    /**
     * @param string      $appId      the Application's ID
     * @param string|null $appUid     the Application's UID
     * @param string      $attemptId  the MessageAttempt's ID
     * @param string      $endpointId the Endpoint's ID
     * @param string|null $eventType  The event type's name
     * @param string|null $msgEventId the Message's UID
     * @param string      $msgId      the Message's ID
     * @param string      $orgId      the Environment's ID
     */
    private function __construct(
        public readonly string $appId,
        public readonly int $attemptCount,
        public readonly \DateTimeImmutable $attemptEnd,
        public readonly string $attemptId,
        public readonly \DateTimeImmutable $attemptStart,
        public readonly string $endpointId,
        public readonly \DateTimeImmutable $msgCreated,
        public readonly string $msgId,
        public readonly string $orgId,
        public readonly int $responseStatusCode,
        public readonly MessageStatus $status,
        public readonly ?string $appUid = null,
        public readonly ?string $eventType = null,
        public readonly ?HttpAttemptTimes $httpTimes = null,
        public readonly ?string $msgEventId = null,
        array $setFields = [],
    ) {
        $this->setFields = $setFields;
    }

    /**
     * Create an instance of MessageAttemptLog with required fields.
     */
    public static function create(
        string $appId,
        int $attemptCount,
        \DateTimeImmutable $attemptEnd,
        string $attemptId,
        \DateTimeImmutable $attemptStart,
        string $endpointId,
        \DateTimeImmutable $msgCreated,
        string $msgId,
        string $orgId,
        int $responseStatusCode,
        MessageStatus $status,
    ): self {
        return new self(
            appId: $appId,
            appUid: null,
            attemptCount: $attemptCount,
            attemptEnd: $attemptEnd,
            attemptId: $attemptId,
            attemptStart: $attemptStart,
            endpointId: $endpointId,
            eventType: null,
            httpTimes: null,
            msgCreated: $msgCreated,
            msgEventId: null,
            msgId: $msgId,
            orgId: $orgId,
            responseStatusCode: $responseStatusCode,
            status: $status,
            setFields: ['appId' => true, 'attemptCount' => true, 'attemptEnd' => true, 'attemptId' => true, 'attemptStart' => true, 'endpointId' => true, 'msgCreated' => true, 'msgId' => true, 'orgId' => true, 'responseStatusCode' => true, 'status' => true]
        );
    }

    public function withAppUid(?string $appUid): self
    {
        $setFields = $this->setFields;
        $setFields['appUid'] = true;

        return new self(
            appId: $this->appId,
            appUid: $appUid,
            attemptCount: $this->attemptCount,
            attemptEnd: $this->attemptEnd,
            attemptId: $this->attemptId,
            attemptStart: $this->attemptStart,
            endpointId: $this->endpointId,
            eventType: $this->eventType,
            httpTimes: $this->httpTimes,
            msgCreated: $this->msgCreated,
            msgEventId: $this->msgEventId,
            msgId: $this->msgId,
            orgId: $this->orgId,
            responseStatusCode: $this->responseStatusCode,
            status: $this->status,
            setFields: $setFields
        );
    }

    public function withEventType(?string $eventType): self
    {
        $setFields = $this->setFields;
        $setFields['eventType'] = true;

        return new self(
            appId: $this->appId,
            appUid: $this->appUid,
            attemptCount: $this->attemptCount,
            attemptEnd: $this->attemptEnd,
            attemptId: $this->attemptId,
            attemptStart: $this->attemptStart,
            endpointId: $this->endpointId,
            eventType: $eventType,
            httpTimes: $this->httpTimes,
            msgCreated: $this->msgCreated,
            msgEventId: $this->msgEventId,
            msgId: $this->msgId,
            orgId: $this->orgId,
            responseStatusCode: $this->responseStatusCode,
            status: $this->status,
            setFields: $setFields
        );
    }

    public function withHttpTimes(?HttpAttemptTimes $httpTimes): self
    {
        $setFields = $this->setFields;
        $setFields['httpTimes'] = true;

        return new self(
            appId: $this->appId,
            appUid: $this->appUid,
            attemptCount: $this->attemptCount,
            attemptEnd: $this->attemptEnd,
            attemptId: $this->attemptId,
            attemptStart: $this->attemptStart,
            endpointId: $this->endpointId,
            eventType: $this->eventType,
            httpTimes: $httpTimes,
            msgCreated: $this->msgCreated,
            msgEventId: $this->msgEventId,
            msgId: $this->msgId,
            orgId: $this->orgId,
            responseStatusCode: $this->responseStatusCode,
            status: $this->status,
            setFields: $setFields
        );
    }

    public function withMsgEventId(?string $msgEventId): self
    {
        $setFields = $this->setFields;
        $setFields['msgEventId'] = true;

        return new self(
            appId: $this->appId,
            appUid: $this->appUid,
            attemptCount: $this->attemptCount,
            attemptEnd: $this->attemptEnd,
            attemptId: $this->attemptId,
            attemptStart: $this->attemptStart,
            endpointId: $this->endpointId,
            eventType: $this->eventType,
            httpTimes: $this->httpTimes,
            msgCreated: $this->msgCreated,
            msgEventId: $msgEventId,
            msgId: $this->msgId,
            orgId: $this->orgId,
            responseStatusCode: $this->responseStatusCode,
            status: $this->status,
            setFields: $setFields
        );
    }

    public function jsonSerialize(): mixed
    {
        $data = [
            'appId' => $this->appId,
            'attemptCount' => $this->attemptCount,
            'attemptEnd' => $this->attemptEnd->format('c'),
            'attemptId' => $this->attemptId,
            'attemptStart' => $this->attemptStart->format('c'),
            'endpointId' => $this->endpointId,
            'msgCreated' => $this->msgCreated->format('c'),
            'msgId' => $this->msgId,
            'orgId' => $this->orgId,
            'responseStatusCode' => $this->responseStatusCode,
            'status' => $this->status];

        if (isset($this->setFields['appUid'])) {
            $data['appUid'] = $this->appUid;
        }
        if (isset($this->setFields['eventType'])) {
            $data['eventType'] = $this->eventType;
        }
        if (isset($this->setFields['httpTimes'])) {
            $data['httpTimes'] = $this->httpTimes;
        }
        if (isset($this->setFields['msgEventId'])) {
            $data['msgEventId'] = $this->msgEventId;
        }

        return \Svix\Utils::newStdClassIfArrayIsEmpty($data);
    }

    /**
     * Create an instance from a mixed obj.
     */
    public static function fromMixed(mixed $data): self
    {
        return new self(
            appId: \Svix\Utils::deserializeString($data, 'appId', true, 'MessageAttemptLog'),
            appUid: \Svix\Utils::deserializeString($data, 'appUid', false, 'MessageAttemptLog'),
            attemptCount: \Svix\Utils::deserializeInt($data, 'attemptCount', true, 'MessageAttemptLog'),
            attemptEnd: \Svix\Utils::deserializeDt($data, 'attemptEnd', true, 'MessageAttemptLog'),
            attemptId: \Svix\Utils::deserializeString($data, 'attemptId', true, 'MessageAttemptLog'),
            attemptStart: \Svix\Utils::deserializeDt($data, 'attemptStart', true, 'MessageAttemptLog'),
            endpointId: \Svix\Utils::deserializeString($data, 'endpointId', true, 'MessageAttemptLog'),
            eventType: \Svix\Utils::deserializeString($data, 'eventType', false, 'MessageAttemptLog'),
            httpTimes: \Svix\Utils::deserializeObject($data, 'httpTimes', false, 'MessageAttemptLog', [HttpAttemptTimes::class, 'fromMixed']),
            msgCreated: \Svix\Utils::deserializeDt($data, 'msgCreated', true, 'MessageAttemptLog'),
            msgEventId: \Svix\Utils::deserializeString($data, 'msgEventId', false, 'MessageAttemptLog'),
            msgId: \Svix\Utils::deserializeString($data, 'msgId', true, 'MessageAttemptLog'),
            orgId: \Svix\Utils::deserializeString($data, 'orgId', true, 'MessageAttemptLog'),
            responseStatusCode: \Svix\Utils::deserializeInt($data, 'responseStatusCode', true, 'MessageAttemptLog'),
            status: \Svix\Utils::deserializeObject($data, 'status', true, 'MessageAttemptLog', [MessageStatus::class, 'fromMixed'])
        );
    }

    /**
     * Create an instance from a json string.
     */
    public static function fromJson(string $json): self
    {
        $data = json_decode(json: $json, associative: true, depth: 512, flags: JSON_THROW_ON_ERROR);

        return self::fromMixed($data);
    }
}
